Replace count_loc.sh with xtask

joshka opened the pull request:
To run the count_loc now, you can use `cargo xtask count-loc` instead of
`./count_loc.sh`.

See <https://github.com/matklad/cargo-xtask> for more info on
cargo-xtask.

Fixes: https://github.com/ilai-deutel/kibi/issues/323

joshka commented:
<img width="732" alt="image" src="https://github.com/user-attachments/assets/0167ce03-8e88-427d-85d7-7a62616b84b9">

ilai-deutel commented on README.md:
````diff
@@ -307,7 +307,14 @@ editor written by Salvatore Sanfilippo (antirez) in C, and
+cargo xtask count-loc
+```
 
 This project must remain tiny, so using advanced dependencies such as [`ncurses`](https://crates.io/crates/ncurses),
````
Actually, could you keep this section below under ## Dependencies? This was more meant as an explanation of the project's philosophy than contributing guidelines

joshka replied:
e16f332..f24b462

ilai-deutel commented on Cargo.toml:
```diff
@@ -1,3 +1,7 @@
+[workspace]
```
Reading the doc, it looks like xtask can be used as part of a workspace or not. For this project I think it would be best to not have it as a workspace, so kibi's `Cargo.lock` can stay clean of dev dependencies, and the xtask's one can be used for dev tools.

What do you think, do you mind changing that? It looks like it just requires changing the alias to `run --manifest-path ./xtask/Cargo.toml --` and removing this `[workspace]` section

joshka replied:
e16f332..f24b462

ilai-deutel replied:
Good point! A workspace sounds good then.

joshka replied:
1407de8..40755f6

ilai-deutel reviewed (commented):
Amazing, thank you so much!

joshka commented on xtask/src/count_loc.rs:
```diff
@@ -0,0 +1,80 @@
+            println!("{:width$} {:4}  {}", header.bold(), total.bold(), note.green());
+        };
+    }
+    if too_high {
+        bail!("Total count is too high");
+    }
```
Yep, sounds good.

joshka replied:
e16f332..f24b462

ilai-deutel replied:
Can the dependency on `color_eyre` be removed then? Returning a native error here would do right?

joshka commented:
Apologies for the rebases here - missed removing the intentional failure code and reverting it back properly (and then got caught by vscode autoformatting)

ilai-deutel commented on xtask/src/count_loc.rs:
```diff
@@ -0,0 +1,80 @@
+    for entry in glob("./*.rs")?.chain(glob("./src/**/*.rs")?) {
+        let path = entry?;
+        let tempfile = filter_lines(&path)?;
+
```
I just looked at the tokei doc, looks like you can avoid writing to temp files by calling `LanguageType::Rust::parse_from_str(...)` to get the `CodeStats` directly

joshka replied:
d322c3c..1407de8

joshka commented:
I'll take a look at the current set of failures a bit later

joshka commented:
Rebased and squashed to re-trigger checks, I think the last failure might have been a transient rust nightly issue rather than a true failure. Re-running to confirm.

joshka commented:
with --package: a different error appears...

```
Run mbrobbel/rustfmt-check@master
  with:
    token: ***
    mode: review
    args: --package kibi
    commit-message: Format Rust code using rustfmt
  env:
    CARGO_TERM_COLOR: always
/home/runner/.cargo/bin/cargo +nightly fmt --package kibi -- --emit json
[]
Error: Resource not accessible by integration
```

joshka commented:
Raised https://github.com/mbrobbel/rustfmt-check/issues/1137 against the rustfmt check action to try to find out more about the problem.

joshka commented:
https://github.com/mbrobbel/rustfmt-check/issues/1137#issuecomment-2478342526

> Looks like this is happening on a PR from a forked repo. Depending on the repo configuration the token may not get the required write permissions, resulting in the `Resource not accessible by integration` error. More info here https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token.
> 
> Can you maybe rerun the failed job and select `Enable debug logging.`?

ilai-deutel commented:
Thank you for raising this with the action owner! I added this yesterday to reduce the friction with fmt checks and it worked when I tested it in a different PR, I'm not sure what's going on. I shared the debug logs in the other issue.

ilai-deutel commented:
I just saw you modified the CI YAML file. I wonder if this is a security feature, disabling token permissions for non-owners who modify the config files to avoid running arbitrary code with all the permissions.

Could you try maybe adding `rustfmt::skip` codeblocks in the xtask code as needed, and revert your change to the yaml?

ilai-deutel commented:
I disabled `mbrobbel/rustfmt-check` in the workflow config for now

joshka commented:
Rebased on master - all green
